app/external_doc.rb

```diff
@@ -59,17 +59,15 @@ def call
 
         href = element["href"].strip
         uri = URI.parse(href)
-        path = uri.path
 
         next if uri.scheme || href.start_with?("#")
 
-        base = if path.start_with? "/"
-                 base_url
-               else
-                 context[:subpage_url]
-               end
-
-        element["href"] = URI.join(base, href).to_s
+        if href.start_with?("/") && context[:subpage_url].to_s.start_with?("https://github.com")
+          # remove preceding "/" to make links relative to the repository
+          # rather than to github.com. This is what GitHub does too.
+          href = href[1..-1]
```

Comment:
```suggestion
```

`URI.join` is agnostic of whether the suffix starts with a `/`.

Reply:
Apparently it's not. If I don't remove the preceding `/`, it converts to `https://github.com/public/foo.json` (arguably correctly).

Reply:
Oh, that's strange. Just to check my understanding:

- `/public/foo.json` -> `https://github.com/public/foo.json`
- `public/foo.json` -> `https://github.com/alphagov/lipsum/blob/master/public/foo.json`

Is that right?

Reply:
That's right. Assuming `context[:subpage_url]` is something like `https://github.com/alphagov/foo`, and the document we're processing is `bar.html`, that would have a path of `https://github.com/alphagov/foo/bar.html`. Any links inside `bar.html` that begin with `/` will be relative to the site root, so `/fish` becomes `https://github.com/fish`.

And that's fine most of the time, but not in GitHub's case where people incorrectly use relative-to-root links :)

Reply:
Wow, that's surprisingly unintuitive - I only tried it with a root path. I'm seeing the same behaviour. Worse, it seems to rely on the `context[:subpage_url]` ending in a `/` 😠.

Would it actually be valid to just do `context[:subpage_url] + href[1..]`? We know the former ends in a `/`, and `href[1..]` doesn't, so I'm not sure what value `URI.join` is adding.

Reply:
It feels fairly intuitive to me - `URI.join` seems to be following the same URL interpretation rules I'd expect in browser. I'm surprised that `context[:subpage_url]` needs to end in a `/`, but on reflection I suppose it makes sense; if you're on `https://google.com/foo` and there's a link to simply `bar`, I'd expect it to take me to `/bar` rather than `/foo/bar`.

We could concatenate the strings manually, though that does risk adding lots of extraneous slashes (or worse, no slashes), depending on the two vars we're joining. I think we ought to leave it as-is.

I hadn't spotted that I could use `href[1..]` instead of `href[1..-1]` - I'll apply that! The StackOverflow I took it from probably assumed it needed to do some array housekeeping or something.

Reply:
> that does risk adding lots of extraneous slashes 

I don't think that's true. My comment here is because we're actually making two assumptions about the components being joined, and that those assumptions would be more clearly expressed with a simple concatenation.

- We're assuming `context[:subpage_url]` ends in a `/`. Otherwise the URL will be incorrect.
- We're assuming `href` begins with a `/`, as stated in the `if` clause.

So we know enough about the structure of the two strings to simply concatenate them. URI.join seems to be more focussed on replicating browser behaviour them helpfully constructing a valid URI, which is really what we're trying to do here. All of this is a relatively minor comment, though, so I agree we shouldn't get hung up on it.

Reply:
Hmm looking at the code again, it is true that we know what strings we're dealing with, so it would normally be safe to simply concatenate them. _However_, `context[:subpage_url]` is of type `URI::HTTP`, rather than a string, so it will still use the `URI.join` concatenation method even if we're concatenating a normal string:

```ruby
foo = URI.join("http://www.google.com", "foo")
=> #<URI::HTTP http://www.google.com/foo>
irb(main):002:0> foo + "bar"
=> #<URI::HTTP http://www.google.com/bar>

# (expected output to be `http://www.google.com/foobar`)
```

Given that we use `URI.join` for every URI creation in this file, I am inclined to leave this as-is.